File: source/phpfiles/utils/weha/categorizer/ActionCategorizer.php
<?php
include_once( dirname(__FILE__) . '/../diff/SentenceDiff.php');
include_once( dirname(__FILE__) . '/../WikiLexer.php');
include_once('EditorialComments.php');
include_once('Categorize.php');
include_once('InterwikiLinks.php');
include_once('Wikify.php');
include_once('Dewikify.php');
include_once('PunctuationCorrection.php');
include_once('ImageAttribute.php');
include_once('TypoCorrection.php');
include_once('References.php');
include_once('ContentRemoval.php');
include_once('ContentAddition.php');
include_once('ContentSubstitution.php');
include_once('ContentMovement.php');
include_once('Uncategorized.php');

class ActionCategorizer {
	private $oldTokens;
	private $newTokens;
	private $basicEdits;
	private $sdiff;
	private $actions;
	
	public function __construct($oldStr, $newStr) {
		$scanner = new WikiLexer($oldStr);
		$this->oldTokens = $scanner->getWikiTokens();
		
		$scanner = new WikiLexer($newStr);
		$this->newTokens = $scanner->getWikiTokens();
		
		$this->sdiff = new SentenceDiff($this->oldTokens, $this->newTokens);
		
		$this->actions = array();
		$this->actions[] = new EditorialComments();
		$this->actions[] = new Categorize();
		$this->actions[] = new InterwikiLinks();
		$this->actions[] = new Wikify();
		$this->actions[] = new Dewikify();
		$this->actions[] = new PunctuationCorrection();
		$this->actions[] = new ImageAttribute();
		$this->actions[] = new TypoCorrection();
		$this->actions[] = new References();
		$this->actions[] = new ContentSubstitution();
		$this->actions[] = new ContentRemoval();
		$this->actions[] = new ContentAddition();
		$this->actions[] = new ContentMovement();
		$this->actions[] = new Uncategorized();
	}

	public function getBasicEdits() {
		return $this->basicEdits;
	}

	public function printResult()
	{
		$retString = array();
		$outStr = "";
		
		$retString[] = $this->sdiff->separateSentences();
		$this->sdiff->exactMatch();
		$retString[] = $this->sdiff->diff();
		
		$this->basicEdits = $this->sdiff->outputDiff();
		$retString[] = $this->sdiff->matchingSentence();
		$retString[] = $this->sdiff->mergeSentences();
		
		for ($i = 0; $i < count($this->basicEdits); $i++)
		{
			if (!($this->basicEdits[$i] instanceof Match))
				$outStr .= $this->basicEdits[$i]->getDescription();
				$outStr .= "\n";
		}
		$outStr .= "\n";
		
		$retString[] = $outStr;
		
		return $retString;
	}
	
	public function categorize()
	{
		$retString = "";
		$beList = $this->basicEdits;
		$ret = array();
		
		for ($j = 0; $j < count($this->actions); $j++) {
			$ae = $this->actions[$j];
			$ret = array_merge($ret, $ae->classify($beList));
		}
		
		// Print the categorized actions 
		foreach ($ret as $action)
		{
			$retString .= get_class($action) . " ";
			for ($i = 0; $i < count($action->be); $i++)
				$retString .= $action->be[$i]->getDescription() . "; ";
			$retString .= "\n";
		}
		
		return $retString;
	}
	
	public function getTokenOld() {
		return $this->oldTokens;
	}
	
	public function getTokenNew() {
		return $this->newTokens;
	}
	
	public function getSentenceEdits() {
		return $this->sdiff->getSentenceEdits();
	}
}

File: weha/source/weha/wikianalysis/categorizer/ActionCategorizer.java
package mo.umac.wikianalysis.categorizer;

import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.Reader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.ListIterator;

import mo.umac.wikianalysis.diff.sentence.SentenceDiff;
import mo.umac.wikianalysis.diff.sentence.SentenceEdit;
import mo.umac.wikianalysis.diff.token.BasicEdit;
import mo.umac.wikianalysis.diff.token.Match;
import mo.umac.wikianalysis.diff.token.Movement;
import mo.umac.wikianalysis.lexer.WikiToken;
import mo.umac.wikianalysis.lexer.WikitextTokenizer;

public class ActionCategorizer {
	private BasicEdit[] basicEdits;
	private String basicEditStr;
	private SentenceDiff sdiff;
	private ArrayList<AbstractEditAction> actions;
	private ArrayList<AbstractEditAction> actionList;
	
	public ActionCategorizer(String oldStr, String newStr) {
		this.sdiff = new SentenceDiff(oldStr, newStr);
		
		this.actions = new ArrayList<AbstractEditAction>();
		this.actions.add(new EditorialComments());
		this.actions.add(new Categorize());
		this.actions.add(new Interwiki());
		this.actions.add(new Wikify());
		this.actions.add(new Dewikify());
		this.actions.add(new PunctuationCorrection());
		this.actions.add(new ImageAttribute());
		this.actions.add(new TypoCorrection());
		this.actions.add(new ImageAddition());
		this.actions.add(new ImageRemoval());
		this.actions.add(new References());
		this.actions.add(new ContentSubstitution());
		this.actions.add(new ContentRemoval());
		this.actions.add(new ContentAddition());
		this.actions.add(new ContentMovement());
		this.actions.add(new Uncategorized());
	}
	
	public ActionCategorizer(WikiToken[] oldT, WikiToken[] newT) {
		this.sdiff = new SentenceDiff(oldT, newT);
		
		this.actions = new ArrayList<AbstractEditAction>();
		this.actions.add(new EditorialComments());
		this.actions.add(new Categorize());
		this.actions.add(new Interwiki());
		this.actions.add(new Wikify());
		this.actions.add(new Dewikify());
		this.actions.add(new PunctuationCorrection());
		this.actions.add(new ImageAttribute());
		this.actions.add(new TypoCorrection());
		this.actions.add(new ImageAddition());
		this.actions.add(new ImageRemoval());
		this.actions.add(new References());
		this.actions.add(new ContentSubstitution());
		this.actions.add(new ContentRemoval());
		this.actions.add(new ContentAddition());
		this.actions.add(new ContentMovement());
		this.actions.add(new Uncategorized());
	}

	public BasicEdit[] getBasicEdits() {
		return basicEdits;
	}

	public String printBasicEdits() {
		if (basicEditStr != null)
			return basicEditStr;
		
		StringBuffer retVal = new StringBuffer();
		
		for (int i = 0; i < basicEdits.length; i++)
			retVal.append(basicEdits[i].getDescription() + "\n");
		
		return retVal.toString();
	}
	
	public String[] printResult()
	{
		String[] retString = new String[4];
		StringBuffer outStr = new StringBuffer();
		
		retString[0] = sdiff.separateSentences();
		sdiff.exactMatch();
		retString[1] = sdiff.diff();
		
		basicEdits = sdiff.outputDiff();
		basicEditStr = this.printBasicEdits();
		retString[2] = sdiff.matchingSentence();
		
		for (int i = 0; i < basicEdits.length; i++)
		{
			if (!(basicEdits[i] instanceof Match) ||
				 (basicEdits[i] instanceof Movement))
				outStr.append(basicEdits[i].getDescription() + "\n");
		}
		outStr.append("\n");
		
		retString[3] = outStr.toString();
		
		return retString;
	}
	
	public List<AbstractEditAction> categorize()
	{
		if (this.actionList == null)
		{
			ArrayList<BasicEdit> beList = new ArrayList<BasicEdit>(Arrays.asList(basicEdits));
			ArrayList<AbstractEditAction> ret = new ArrayList<AbstractEditAction>();
			
			for (int j = 0; j < actions.size(); j++) {
				AbstractEditAction ae = actions.get(j);
				ret.addAll(ae.classify(beList));
			}
			
			this.actionList = ret;
		}
		return this.actionList;
	}
	
	public String printCategorize()
	{
		StringBuffer retString = new StringBuffer();
		
		// Print the categorized actions 
		ListIterator<AbstractEditAction> iter = actionList.listIterator();
		while (iter.hasNext())
		{
			AbstractEditAction action = iter.next();
			retString.append(action.getClass().getSimpleName() + " ");
			for (int i = 0; i < action.be.length; i++)
				retString.append(action.be[i].getDescription() + "; ");
			retString.append("\n");
		}
		
		return retString.toString();
	}
	
	public String[] printLinkedCategorize()
	{
		ArrayList<String> catList = new ArrayList<String>();
		String prevName = new String();
		StringBuffer retString = new StringBuffer();
		
		// Print the categorized actions 
		ListIterator<AbstractEditAction> iter = actionList.listIterator();
		while (iter.hasNext())
		{
			AbstractEditAction action = iter.next();
			String curName = action.getClass().getSimpleName();
			if (!curName.equals(prevName)) {
				catList.add(prevName + "\n" + retString.toString());
				retString = new StringBuffer();
			}
			for (int i = 0; i < action.be.length; i++)
				retString.append(action.be[i].getLinkedDesc() + "; ");
			retString.append("\n");

			prevName = curName;
		}
		
		catList.add(prevName + "\n" + retString.toString());
		String[] catArray = new String[0];
		catArray = catList.toArray(catArray);
		
		return catArray;
	}
	
	public WikiToken[] getTokenOld() {
		return sdiff.getOldTokens();
	}
	
	public WikiToken[] getTokenNew() {
		return sdiff.getNewTokens();
	}
	
	public List<SentenceEdit> getSentenceEdits() {
		return sdiff.getSentenceEdits();
	}
	
	public String printSentenceEdits() {
		return sdiff.printSentenceEdits();
	}
	
	public static void main(String [] args)
	{
		StringBuffer oldTextBuf = new StringBuffer();
		StringBuffer newTextBuf = new StringBuffer();
		String tmpText;
		
		try {
     		Reader oldIn = new InputStreamReader(new FileInputStream("/home/peter/Desktop/Examples/oldText.txt"), "UTF-8");
			BufferedReader oldFile = new BufferedReader(oldIn);
			
			do {
				tmpText = oldFile.readLine();
				if (tmpText != null)
				{
					oldTextBuf.append(tmpText);
					oldTextBuf.append("\n");
				}
			} while (tmpText != null);
			
			oldFile.close();
			
     		Reader newIn = new InputStreamReader(new FileInputStream("/home/peter/Desktop/Examples/newText.txt"), "UTF-8");
			BufferedReader newFile = new BufferedReader(newIn);
			
			do {
				tmpText = newFile.readLine();
				if (tmpText != null)
				{
					newTextBuf.append(tmpText);
					newTextBuf.append("\n");
				}
			} while (tmpText != null);
			
			newFile.close();
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		} catch (IOException e) {
			e.printStackTrace();
		}
		
		WikiToken[] to = WikitextTokenizer.tokenize(oldTextBuf.toString());
		WikiToken[] tn = WikitextTokenizer.tokenize(newTextBuf.toString());
		
		ActionCategorizer acter = new ActionCategorizer(to, tn);
		
		String[] result = acter.printResult();
		for (int i = 0; i < result.length; i++)
			System.out.println(result[i]);
		
		acter.categorize();
		System.out.println(acter.printCategorize());

		System.out.println(acter.printBasicEdits());
		System.out.println(acter.printSentenceEdits());
	}
	
}
